ZFSin/zfs/module/zfs/zfs_windows_zvol_scsi.c

```diff
@@ -642,52 +600,38 @@ ScsiOpModeSense(
 UCHAR
 ScsiOpReportLuns(                                     
                  __in __out pHW_HBA_EXT         pHBAExt,   // Adapter device-object extension from StorPort.
-                 __in       pHW_LU_EXTENSION    pLUExt,    // LUN device-object extension from StorPort.
                  __in       PSCSI_REQUEST_BLOCK pSrb
                 )
 {
     UCHAR     status = SRB_STATUS_SUCCESS;
     PLUN_LIST pLunList = (PLUN_LIST)pSrb->DataBuffer; // Point to LUN list.
-    ULONG     i, 
-              GoodLunIdx;
-
-    UNREFERENCED_PARAMETER(pLUExt);
+    uint8_t   GoodLunIdx = 0;
+	uint8_t   totalLun = 0;
 
     if (FALSE==pHBAExt->bReportAdapterDone) {         // This opcode will be one of the earliest I/O requests for a new HBA (and may be received later, too).
         wzvol_HwReportAdapter(pHBAExt);                   // WMIEvent test.
-
         wzvol_HwReportLink(pHBAExt);                      // WMIEvent test.
-
         wzvol_HwReportLog(pHBAExt);                       // WMIEvent test.
-
         pHBAExt->bReportAdapterDone = TRUE;
     }
 	
-	zvol_state_t *zv;
-
-	zv = wzvol_find_target(pSrb->TargetId, pSrb->Lun);
-
-	if (zv != NULL &&
-		pSrb->PathId == 0 &&
-		!pHBAExt->bDontReport) {
-
-		RtlZeroMemory((PUCHAR)pSrb->DataBuffer, pSrb->DataTransferLength);
-
-        pLunList->LunListLength[3] =                  // Set length needed for LUNs.
-            (UCHAR)(8*pHBAExt->NbrLUNsperHBA);
-
-        // Set the LUN numbers if there is enough room, and set only those LUNs to be reported.
-
-        if (pSrb->DataTransferLength>=FIELD_OFFSET(LUN_LIST, Lun) + (sizeof(pLunList->Lun[0])*pHBAExt->NbrLUNsperHBA)) {
-            for (i = 0, GoodLunIdx = 0; i < pHBAExt->NbrLUNsperHBA; i ++) {
-                // LUN to be reported?
-                if (FALSE==pHBAExt->LUNInfoArray[i].bReportLUNsDontUse) {
-                    pLunList->Lun[GoodLunIdx][1] = (UCHAR)i;
-                    GoodLunIdx++;
-                }
-            }
+	RtlZeroMemory((PUCHAR)pSrb->DataBuffer, pSrb->DataTransferLength);
+	if (!pHBAExt->bDontReport) {
+        // Set the LUN numbers if there is enough room, and set only those LUNs to be reported.       
+        for (uint8_t i = 0; i < STOR_wzvolDriverInfo.MaximumNumberOfLogicalUnits; i ++) {
+			// make sure we have the space for 1 more LUN each time.
+			if (wzvol_find_target(pSrb->TargetId, i)) {
+				totalLun++;
+				if (pSrb->DataTransferLength >= FIELD_OFFSET(LUN_LIST, Lun) + (GoodLunIdx * sizeof(pLunList->Lun[0])) + sizeof(pLunList->Lun[0])) {
+					pLunList->Lun[GoodLunIdx][1] = (UCHAR)i;
+					GoodLunIdx++;
+				}
+			}
         }
-    }
+    } // else: we chose to not report any LUN through that HBA (see FindAdapter routine).
+
+	*((ULONG*)&pLunList->LunListLength) = RtlUlongByteSwap(totalLun * sizeof(pLunList->Lun[0]));
```

Comment:
Actually, not sure why this looks so complicated, a simple
`pLunList->LunListLength = RtlUlongByteSwap(totalLun * sizeof(pLunList->Lun[0]));`
should work? Is LunListLength not ULONG ?

Reply:
` UCHAR LunListLength[4];`
Ah! and ew! :)  Ok, that makes sense.